misc/python/materialize/cli/scratch/destroy.py

```diff
@@ -8,26 +8,73 @@
 # by the Apache License, Version 2.0.
 
 import argparse
+import sys
+from typing import List
 
 import boto3
+from mypy_boto3_ec2.type_defs import FilterTypeDef
 
 from materialize.cli.scratch import check_required_vars
-from materialize.scratch import print_instances
+from materialize.scratch import print_instances, whoami
 
 
 def configure_parser(parser: argparse.ArgumentParser) -> None:
     parser.add_argument(
         "instances",
-        nargs="+",
+        nargs="*",
         help="Instance IDs to destroy",
     )
+    parser.add_argument(
+        "--all-mine",
+        action="store_true",
+        help="Destroy all of your instances (incompatible with specifying instance IDs)",
+    )
+    parser.add_argument(
+        "-n",
+        "--dry-run",
+        action="store_true",
+        help="Don't actually destroy any instances",
+    )
     parser.add_argument("--output-format", choices=["table", "csv"], default="table")
 
 
 def run(args: argparse.Namespace) -> None:
     check_required_vars()
-    ists = list(boto3.resource("ec2").instances.filter(InstanceIds=args.instances))
-    for i in ists:
-        i.terminate()
-    print("Destroyed instances:")
-    print_instances(ists, args.output_format)
+    instance_ids = []
+    filters: List[FilterTypeDef] = [
+        {
+            "Name": "instance-state-name",
+            "Values": ["pending", "running", "shutting-down", "stopping", "stopped"],
+        }
+    ]
+    if args.all_mine:
+        if args.instances:
+            print(
+                "scratch: error: cannot specify --all-mine and instance IDs",
+                file=sys.stderr,
+            )
+            sys.exit(1)
+        filters.append({"Name": "tag:LaunchedBy", "Values": [whoami()]})
+    elif not args.instances:
+        print(
+            "scratch: error: must supply at least one instance ID to destroy",
+            file=sys.stderr,
+        )
```

Comment:
I think `argparse` can actually express this constraint with https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_mutually_exclusive_group, if you want to clean up this code

Reply:
It doesn't work when one of the mutually-exclusive things is a positional argument unfortunately :(